internal/internal.go

```diff
@@ -153,14 +155,52 @@ func (p *envoyExtAuthzGrpcServer) Check(ctx ctx.Context, req *ext_authz.CheckReq
 		return nil, err
 	}
 
-	status := int32(google_rpc.PERMISSION_DENIED)
+	resp := &ext_authz.CheckResponse{}
 
-	if result.decision {
-		status = int32(google_rpc.OK)
-	}
+	switch decision := result.decision.(type) {
+	case bool:
+		status := int32(google_rpc.PERMISSION_DENIED)
+		if decision {
+			status = int32(google_rpc.OK)
+		}
+
+		resp.Status = &google_rpc.Status{Code: status}
+
+	case map[string]interface{}:
+		status, err := getResponseStatus(decision)
+		if err != nil {
+			return nil, err
```

Comment:
I'd consider wrapping the error so that we can determine whether it was the status or the headers or the body that caused the error.

Reply:
Wrapped errors to include more context.